[2.3] Transfer Encoding of emails changed to QUOTED-PRINTABLE

gwharton opened the pull request:
### Description (*)
When Magento used Zend Framework 1 for emails, MIME emails were sent with Transfer Encoding set to QUOTED-PRINTABLE. This means that emails with long lines (particularly emails with large inline css) are encoded properly. This can be confirmed by running tests under 2.2.7.

When Magento switched to Zend Framework 2, the default Transfer Encoding was set to 8-bit. This emails with long lines were being rejected by some mail servers as they were not encoded quoted-printable as expected.

Magento 2.2.8+ and 2.3.0+ are affected by this.

This PR changes the default encoding of all MIME emails to quoted-printable.

### Fixed Issues (if relevant)
1. magento/magento2#23643: Mime parts of email are no more encoded with quoted printable

### Manual testing scenarios (*)
```
Backend
=======
Marketing -> Email Templates -> Add New Template
    Template Name = Newsletter Success Plain
    Subject = Success Subject
    Template Content = Success Body
    Convert to Plain Text
    Save Template
Marketing -> Email Templates -> Add New Template
    Template Name = Newsletter Success Html
    Subject = Success Subject
    Template Content = <html><body>Success Body</body></html>
    Save Template
Store -> Configuration -> General -> Store Email Addresses
    General Contact
        Sender Name = Owner Name
        Sender Email = owner@domain.com
Store -> Configuration -> Customers -> Newsletter -> Subscription Options
    Success Email Template = Newsletter Success Plain

Frontend
========
Subscribe to newsletter

Ensure email received contains the following headers

MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: quoted-printable

Backend
=======
Store -> Configuration -> Customers -> Newsletter -> Subscription Options
    Success Email Template = Newsletter Success Html

Frontend
========
Subscribe to newsletter

Ensure email received contains the following headers

MIME-Version: 1.0
Content-Type: text/html; charset="utf-8"
Content-Transfer-Encoding: quoted-printable
```

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

pmclain reviewed (dismissed):
@gwharton Please review the integration test checks https://testing-service.magento-community.engineering/reports/magento/magento2/pull/23649/948900ad-78aa-453d-b0f1-ee38478a911e/11125/Integration/allure-report-ce/index.html

pmclain commented:
I'm also struggling to find where earlier versions used quoted-printable for the message body. The magento fork for ZF1 looks to have the [same default](https://github.com/magento/zf1/blob/master/library/Zend/Mime/Part.php#L50) as what is currently used.

ZF1 was using quoted-printable for [encoding headers](https://github.com/magento/zf1/blob/master/library/Zend/Mail.php#L92) and that does seem to be missing from the current ZF2 implementation.

gwharton commented:
Take a look at the ZF1 Zend_Mail class here.

https://github.com/magento/zf1/blob/master/library/Zend/Mail.php#L362-L377

and

https://github.com/magento/zf1/blob/master/library/Zend/Mail.php#L404-L419

which are the setBodyText and setBodyHtml functions from ZF1. In Magento 2.2.7, /Magento/Framework/Mail/Message.php inherited from Zend_Mail. 

With the move to ZF2, these two functions were effectively moved into /Magento/Framework/Mail/Message.php. 

It could also be argued that Zend_Mime::DISPOSITION_INLINE should also be set, as that was also the default in ZF1.

gwharton commented:
With request to the test failures, I'm not at all sure what to do. The tests are failing because encoding the email with quoted-printable, results in an email which has the body split into multiple lines with CRLF at the end of each line. Thus tests that are looking for strings in the email body fail because in many cases the string that is being looked for has a CRLF inserted into it, if it spans a line boundary. I'll see if I can find out why these tests were passing under 2.2 which did encode emails in this way.

gwharton commented:
@magento run all tests

gwharton commented:
OK, i've had to modify the failing integration tests to get them to pass. They were originally written to use the \Zend\Message::getRawMessage function to get the body of the email message, but this function returns the email body encoded. Thus now that the emails are changed to encoding quoted-printable, the raw message returned in the test is encoded and the tests fail.

The tests have been modified to use the \Zend\Mime\Part::getRawContent() instead which returns the email body unencoded, allowing the tests to pass.

pmclain commented on lib/internal/Magento/Framework/Mail/Message.php:
```diff
@@ -167,6 +167,7 @@ private function createMimeFromString($body, $messageType)
     {
         $part = new Part($body);
         $part->setCharset($this->zendMessage->getEncoding());
+        $part->setEncoding(Mime::ENCODING_QUOTEDPRINTABLE);
```
```suggestion
        $part->setEncoding(Mime::ENCODING_QUOTEDPRINTABLE);
        $part->setDisposition(Mime::DISPOSITION_INLINE);
```
I agree with your [comment](https://github.com/magento/magento2/pull/23649#issuecomment-510057282), adding the disposition to match the previous [ZF1 implementation](https://github.com/magento/zf1/blob/master/library/Zend/Mail.php#L413) should be done.

gwharton replied:
Added as requested

pmclain reviewed (commented):
Let's set the disposition to `\Zend\Mime::DISPOSITION_INLINE`. I took a deeper look at the ZF2 header encoding and they are [already quoted-printable](https://github.com/zendframework/zend-mail/blob/master/src/Header/HeaderWrap.php#L91) so we should be good there.

Commit:
Change Encoding of MIME Emails to QUOTED-PRINTABLE
Added Content-Disposition: inline to all MIME emails

Implementation now matches previous ZF1 implementation

gwharton commented:
Backport is here #23650

magento-engcom-team commented:
Hi @pmclain, thank you for the review.
[ENGCOM-5433](https://jira.corp.magento.com/browse/ENGCOM-5433) has been created to process this Pull Request
:eight_spoked_asterisk: @pmclain, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

engcom-Delta commented:
:heavy_check_mark: QA passed
